### PR TITLE
more SonarLint cleanups

### PR DESCRIPTION
**Pull request by neheb**

### CODE REVIEW

**Comment by neheb** [comment]
@kevinbackhouse what do you think about the if constexpr change?

File: src/convert.cpp [hunk 1]
```diff
@@ -274,8 +274,8 @@ class Converter {
 
   // DATA
   static const Conversion conversion_[];  //<! Conversion rules
-  bool erase_;
-  bool overwrite_;
+  bool erase_{};
```
Comment (piponazo) [inline src/convert.cpp:277]: Why did you remove the value `false` here? 

By the way, you could also initialize here the pointers to `nullptr`

File: src/convert.cpp [hunk 2]
```diff
@@ -626,7 +616,12 @@ void Converter::cnvExifDate(const char* from, const char* to) {
     return;
   if (!prepareXmpTarget(to))
     return;
-  int year = 0, month = 0, day = 0, hour = 0, min = 0, sec = 0;
+  int year = 0;
```
Comment (piponazo) [inline src/convert.cpp:619]: I see that you normally use the initialization with curly brackets : `int year{0}`. Why not here?

**Review by piponazo — CHANGES_REQUESTED** [review]
In general the changes look good. I just made some comments and questions before approving

File: src/convert.cpp (hunk 2 above)
Comment (neheb) [reply src/convert.cpp:619]: That's done in member variables

File: src/convert.cpp (hunk 1 above)
Comment (neheb) [reply src/convert.cpp:277]: Constructors already handle that.

File: src/basicio.cpp [hunk 3]
```diff
@@ -159,7 +159,11 @@ int FileIo::Impl::switchMode(OpMode opMode) {
   }
 
   // Reopen the file
-  long offset = std::ftell(fp_);
+#ifdef _WIN32
```
[the remark this answers is not in the document]
Comment (neheb) [reply src/basicio.cpp:162]: IIRC SonarLint was complaining about implicit integer conversions somewhere.

File: src/safe_op.hpp [hunk 4]
```diff
@@ -85,17 +97,6 @@ std::enable_if_t<std::is_signed_v<T> && sizeof(T) >= sizeof(int), bool> fallback
  * Further information:
```
[the remark this answers is not in the document]
Comment (neheb) [reply src/safe_op.hpp:97]: Probably

File: src/sonymn_int.cpp [hunk 5]
```diff
@@ -635,109 +635,75 @@ constexpr TagDetails sonyLongExposureNoiseReduction[] = {
...
+    if (!temp.empty())
+      return temp;
   }
-  val = "";
-  return false;
+  return "";
 }
 
-static auto getFocusMode2(const ExifData* metadata, uint32_t& val) {
+static uint32_t getFocusMode2(const ExifData* metadata) {
```
[the remark this answers is not in the document]
Comment (neheb) [reply src/sonymn_int.cpp:689]: No. I just assumed there are tests for this.

File: src/safe_op.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (neheb) [reply src/safe_op.hpp:97]: Removed.

File: src/convert.cpp (hunk 1 above)
Comment (piponazo) [reply src/convert.cpp:277]: What I meant is that before we had in the constructors this:

```cpp
Converter::Converter(ExifData& exifData, XmpData& xmpData) :
    erase_(false),
    overwrite_(true),
```

And now you moved the initialization to the .h (this for me is fine) where the variables are declared:

```cpp
  bool erase_{};
  bool overwrite_{true};
```

However, you are initializing with an explicit value only `overwrite_` and not `erase_`. Did you forget to assign the `false` value for erase?

Comment (neheb) [reply src/convert.cpp:277]: No. {} and {false} are equivalent. The former is shorter

Comment (piponazo) [reply src/convert.cpp:277]: I do not think so. I have been bitten by such assumptions already many times. Depending on compiler, compiler version, compilation mode (Release or Debug) , platforms, etc. you can get different values when you are not explicit.

Even in the case this changed with latest standards, it is an obscure implementation detail. It is much better to be explicit about your intentions.

**Comment by kevinbackhouse** [comment]
> @kevinbackhouse what do you think about the if constexpr change?

@neheb: That's very cool. I never saw that C++ feature before.

File: src/convert.cpp (hunk 1 above)
Comment (piponazo) [reply src/convert.cpp]: Thanks ;)

**Review by piponazo — DISMISSED** [review]

File: src/crwimage_int.hpp [hunk 6]
```diff
@@ -20,7 +20,7 @@ namespace Exiv2::Internal {
 class CiffHeader;
 class CiffComponent;
 struct CrwMapping;
-using CrwSubDir = struct {
+struct CrwSubDir {
   uint16_t dir;
```
Comment (neheb) [inline src/crwimage_int.hpp:24]: Unbelievable. This section actually fixes compilation on 32-bit ARM on Debian 11 with GCC 9 and 10. I have no idea why. The error is

```
../src/crwimage_int.hpp:239:16: warning: ‘virtual void Exiv2::Internal::CiffComponent::doRemove(Exiv2::Internal::CrwDirs&, uint16_t)’ used but never defined
  239 |   virtual void doRemove(CrwDirs& crwDirs, uint16_t crwTagId);
      |                ^~~~~~~~
../src/crwimage_int.hpp:237:26: warning: ‘virtual Exiv2::Internal::CiffComponent* Exiv2::Internal::CiffComponent::doAdd(Exiv2::Internal::CrwDirs&, uint16_t)’ used but never defined
```

That is just weird

File: src/safe_op.hpp [hunk 7]
```diff
@@ -223,11 +183,13 @@ T add(T summand_1, T summand_2) {
...
-  if (num == std::numeric_limits<T>::min()) {
-    return std::numeric_limits<T>::max();
+T abs(T num) noexcept {
+  if constexpr (std::is_signed_v<T>) {
+    if (num == std::numeric_limits<T>::min())
+      return std::numeric_limits<T>::max();
+    return num < 0 ? -num : num;
   }
-  return num < 0 ? -num : num;
+  return std::abs(num);
```
Comment (kevinbackhouse) [inline src/safe_op.hpp:192]: The type isn't signed here so I don't think there's any need to use `std::abs`.

```suggestion
  return num;
```

File: src/sonymn_int.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinbackhouse) [reply src/sonymn_int.cpp:689]: I agree with @piponazo: I'd prefer to keep the separate Boolean return value to indicate success on all of these functions, to avoid the potential confusion if the field contains `""` or `0`.

[the remark this answers is not in the document]
Comment (neheb) [reply src/sonymn_int.cpp]: removed.

**Review by kevinbackhouse — APPROVED** [review]
